ring/sampler.go

```diff
@@ -1,13 +1,81 @@
 package ring
 
 import (
-	"github.com/tuneinsight/lattigo/v4/utils"
+	"encoding/json"
+	"fmt"
+
+	"github.com/tuneinsight/lattigo/v4/utils/sampling"
+)
+
+const (
+	discreteGaussianName = "DiscreteGaussian"
+	ternaryDistName      = "Ternary"
+	uniformDistName      = "Uniform"
 )
 
-const precision = uint64(56)
+// Sampler is an interface for random polynomial samplers.
+// It has a single Read method which takes as argument the polynomial to be
+// populated according to the Sampler's distribution.
+type Sampler interface {
+	Read(pol Poly)
+	ReadNew() (pol Poly)
+	ReadAndAdd(pol Poly)
+	AtLevel(level int) Sampler
+}
+
+// DistributionParameters is an interface for distribution
+// parameters in the ring.
+// There are three implementation of this interface:
+//   - DiscreteGaussian for sampling polynomials with discretized
+//     gaussian coefficient of given standard deviation and bound.
+//   - Ternary for sampling polynomials with coefficients in [-1, 1].
+//   - Uniform for sampling polynomial with uniformly random
+//     coefficients in the ring.
+type DistributionParameters interface {
+	// Type returns a string representation of the distribution name.
+	Type() string
+	mustBeDist()
```

Comment:
Is the goal here to avoid external implementation of this interface ?

Reply:
Correct. I has a two fold goal: force the user to use this interface & prevent the user from providing custom struct complying to this interface